runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java

```diff
@@ -281,28 +281,43 @@ void registerTable(@Nonnull String namespace,
+     * @param allTableDescriptors  - A map of all the names of the protobuf files with their
+     *                               protobuf file descriptors.
+     * @return true if a schema change was detected and new schema was updated, false otherwise
      */
-    private void recordNewSchema(ProtobufFileName protoName,
-                                    CorfuRecord<ProtobufFileDescriptor, TableMetadata> newProtoFd) {
-        final CorfuRecord<ProtobufFileDescriptor, TableMetadata> currentSchema = this.protobufDescriptorTable.get(protoName);
-        if (currentSchema == null) {
-            // If schema is not present, add to protobufDescriptorTable, otherwise,
-            // we assume schema definitions only change between upgrades/migration for which
-            // dedicated APIs are available
-            this.protobufDescriptorTable.put(protoName, newProtoFd);
-        } else {
-            if (log.isTraceEnabled() && !protoName.getFileName().startsWith("google/protobuf")
-                    && !currentSchema.getPayload().getFileDescriptor()
+    private boolean tryUpdateTableSchemas(Map<ProtobufFileName,
+                                          CorfuRecord<ProtobufFileDescriptor, TableMetadata>> allTableDescriptors) {
+        boolean schemaChangeDetected = false;
+        for (Map.Entry<ProtobufFileName, CorfuRecord<ProtobufFileDescriptor, TableMetadata>> e :
+                allTableDescriptors.entrySet()) {
+            ProtobufFileName protoName = e.getKey();
+            CorfuRecord<ProtobufFileDescriptor, TableMetadata> newProtoFd = e.getValue();
+            final CorfuRecord<ProtobufFileDescriptor, TableMetadata> currentSchema =
+                    this.protobufDescriptorTable.get(protoName);
+            /** Known bug: Protobuf FileDescriptor maps are not deterministically constructed **
+            if (log.isTraceEnabled() && currentSchema != null &&
+                    !protoName.getFileName().startsWith("google/protobuf") &&
+                    !currentSchema.getPayload().getFileDescriptor()
                     .equals(newProtoFd.getPayload().getFileDescriptor())) {
                 log.trace("registerTable: Schema update detected for table {}! " +
                                 "Old schema is {}, new schema is {}", protoName,
                         currentSchema.getPayload().getFileDescriptor(),
                         newProtoFd.getPayload().getFileDescriptor());
-
+            } Uncomment and run once the above bug is either fixed or has a workaround */
+            if (currentSchema != null &&
+                    currentSchema.getPayload().getVersion() == newProtoFd.getPayload().getVersion()) {
+                continue; // old schema is same as the new schema, avoid doing an expensive I/O
+            } // else this process is running a new code version, conservatively update the schemas
+            schemaChangeDetected = true;
```

Comment:
Here we know there is a protobuf definition change (schema evolution) which we want to support based on protobuf's build-in backward-compatibility feature. But the compatibility of this schema evolution has not been validated and here if we just allow to overwrite protobufDescriptorTable, subsequent reads could fail and it's hard to locate the issue at that time. Can we add a validation here which compares if all new fields are appended to the end and the order of existing fields has not changed? If the validation fails then we just fail the openTable.

Reply:
Firstly subsequent reads should not fail if the caller knows what type the read value is. For example during upgrades. During custom migrations, the code commonly reads old data, re-writes it in new and possibly incompatible form violating schema provisions, then deletes the old record. This works today only and we can't break it right?
Secondly, that validation might be good to have in the logs but I don't know of a way to do it using FileDescriptorProto - the constructed map can be a deep tree with a mix of system and custom types. How would we even know if there are appends, safe-type changes (fixed64->int64 for example) and other non-breaking changes that we should not throw error on?

Reply:
The first point is absolutely correct. We are faced with two kinds of schema changes, one is from custom migration which we definitely allow overwrite and there is nothing to validate, the other is 'silent' migration which doesn't have custom migration task and it's up to protobuf's build-in backward-compatibility, and that's where I think we should do some validation. 
I see this can be a non-trivial effort to check compatibility. There is some work outside that does static protobuf file check but that is out of the scope of this PR.

Reply:
https://protolock.dev/

Reply:
thank you @Lujie1996 !